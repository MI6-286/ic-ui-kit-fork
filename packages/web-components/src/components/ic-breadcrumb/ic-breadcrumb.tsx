import {
  Component,
  Host,
  Prop,
  h,
  Element,
  Method,
  Watch,
} from "@stencil/core";
import { IcBreadcrumbDefault } from "./ic-breadcrumb.types";

import chevronIcon from "../../assets/chevron-icon.svg";
import backIcon from "../../assets/back-icon.svg";
import {
  getSlotElements,
  isPropDefined,
  isSlotUsed,
} from "../../utils/helpers";
import { IcThemeMode } from "../../utils/types";

/**
 * @slot icon - Content will be rendered to the left of the breadcrumb page title.
 */
@Component({
  tag: "ic-breadcrumb",
  styleUrl: "ic-breadcrumb.css",
  shadow: {
    delegatesFocus: true,
  },
})
export class Breadcrumb {
  private HREF_ATTR = "href";
  private linkSlotContent: HTMLElement;
  private slottedLinkEl: HTMLElement | null;
  private slottedLinkHref: string | null | undefined;

  @Element() el: HTMLIcBreadcrumbElement;

  /**
   * If `true`, aria-current will be set on the breadcrumb.
   */
  @Prop() current?: boolean = false;
  @Watch("current")
  watchCurrentHandler(): void {
    this.updatedSlottedLinkFocus();
    this.el.ariaCurrent = this.current ? "page" : null;
  }

  /**
   * The URL that the breadcrumb link points to.
   */
  @Prop() href?: string;

  /**
   * @internal If `true`, the breadcrumb will display as black in the light theme, and white in the dark theme.
   */
  @Prop() monochrome?: boolean = false;

  /**
   * The title of the breadcrumb.
   */
  @Prop() pageTitle!: string;

  /**
   * @internal If `true`, back icon will be displayed.
   */
  @Prop({ reflect: true }) showBackIcon?: boolean = false;

  /**
   * @internal Sets the theme color to the dark or light theme color. "inherit" will set the color based on the system settings or ic-theme component.
   */
  @Prop() theme?: IcThemeMode = "inherit";

  // Prevent focus on breadcrumb if current page and contains slotted link
  private updatedSlottedLinkFocus = (): void => {
    // Sets tabindex on wrong element in unit test snapshots
    // - related to known Jest issue: https://github.com/ionic-team/stencil/issues/2830
    if (this.linkSlotContent) {
      this.linkSlotContent.tabIndex = this.current ? -1 : 0; // Prevent focus
    }

    if (this.slottedLinkEl) {
      if (this.current) {
        this.slottedLinkEl.removeAttribute(this.HREF_ATTR); // Prevent screen reader announcing breadcrumb as a link
      } else {
        this.slottedLinkHref &&
          this.slottedLinkEl.setAttribute(this.HREF_ATTR, this.slottedLinkHref);
      }
    }
  };

  private getSlottedLinkEl = (): HTMLElement | null => {
    const link = this.linkSlotContent;
    if (link) {
      const elWithHref = link.hasAttribute(this.HREF_ATTR)
        ? link
        : link.querySelector("[href]");
      return elWithHref as HTMLElement;
    }
    return null;
  };

  componentDidLoad(): void {
    const slottedLinkWrapper =
      this.el.shadowRoot?.querySelector(".link-wrapper");

    if (slottedLinkWrapper) {
      const slotEls = getSlotElements(slottedLinkWrapper);
      if (slotEls) {
        this.linkSlotContent = slotEls[0] as HTMLElement;
      }
      this.slottedLinkEl = this.getSlottedLinkEl();
      this.slottedLinkHref = this.slottedLinkEl?.getAttribute("href");
      this.updatedSlottedLinkFocus();
    }
  }

  componentWillRender(): void {
    this.setSlottedCurrentPageClass();
  }

  /**
   * Sets focus on the breadcrumb.
   */
  @Method()
  async setFocus(): Promise<void> {
    this.el.shadowRoot?.querySelector("ic-link")?.focus();
  }

  private renderBackIcon = () => (
    <div class="back-icon" innerHTML={backIcon}></div>
  );

  private renderDefaultBreadcrumb = (
    current: boolean,
    pageTitle: string,
    describedById: string,
    href: string | undefined
  ): IcBreadcrumbDefault => {
    const hasPageTitle =
      pageTitle !== null && isPropDefined(pageTitle) && pageTitle !== "";

    if (current && hasPageTitle) {
      return (
        <span
          class={{
            "current-page-container": current,
          }}
        >
          {isSlotUsed(this.el, "icon") && <slot name="icon"></slot>}
          {pageTitle}
        </span>
      );
    }

    return (
      <ic-link
        theme={this.theme}
        monochrome={this.monochrome}
        href={href}
        class="breadcrumb-link"
        aria-describedby={this.showBackIcon && describedById && describedById}
      >
        {this.showBackIcon && this.renderBackIcon()}
        {isSlotUsed(this.el, "icon") && <slot name="icon"></slot>}
        {pageTitle}
      </ic-link>
    );
  };

  private setSlottedCurrentPageClass = () => {
    const icLink = this.el.querySelector("ic-link");
    const currentPage = "current-page";
    if (icLink) {
      icLink.classList.remove(currentPage);
      if (this.current) {
        const hasCurrentPageClass = icLink.classList.contains(currentPage);
        if (!hasCurrentPageClass) {
          icLink.classList.add("breadcrumb-link", currentPage);
        }
      }
    }
  };

  render() {
    const { current, href, pageTitle } = this;
    const describedById = `${
      pageTitle && pageTitle.toLowerCase().replace(" ", "-")
    }-describedby`;

    const hasPageTitle =
      pageTitle !== null && isPropDefined(pageTitle) && pageTitle !== "";
    const hasHref = href !== null && isPropDefined(href) && href !== "";

    return (
      <Host
        class={{
          "ic-breadcrumb-back": !!this.showBackIcon,
          [`ic-theme-${this.theme}`]: this.theme !== "inherit",
          "ic-breadcrumb-monochrome": !!this.monochrome,
        }}
        aria-current={current && "page"}
        role="listitem"
      >
        <div class="breadcrumb">
          <span innerHTML={chevronIcon} class="chevron" aria-hidden="true" />
          {this.showBackIcon && describedById && (
            <span
              id={describedById}
              class="hide"
            >{`Back to ${pageTitle}`}</span>
          )}
          {hasPageTitle && hasHref ? (
            this.renderDefaultBreadcrumb(
              !!current,
              pageTitle,
              describedById,
              href
            )
          ) : (
            <div class="slotted-link-container">
              {this.showBackIcon && this.renderBackIcon()}
              <span class="link-wrapper">
                <slot />
              </span>
            </div>
          )}
        </div>
      </Host>
    );
  }
}
