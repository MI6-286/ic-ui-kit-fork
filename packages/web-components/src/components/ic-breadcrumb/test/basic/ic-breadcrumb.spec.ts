import { Breadcrumb } from "../../ic-breadcrumb";
import { newSpecPage } from "@stencil/core/testing";

describe("ic-breadcrumb", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: `<ic-breadcrumb></ic-breadcrumb>`,
    });

    expect(page.root).toMatchSnapshot(`
      <ic-breadcrumb role="listitem">
        <mock:shadow-root>
          <div class="breadcrumb">
            <span aria-hidden="true" class="chevron">
              svg
            </span>
            <span class="link-wrapper" tabindex="0">
              <slot></slot>
            </span>
          </div>
        </mock:shadow-root>
      </ic-breadcrumb>
    `);
  });

  it("should render with page-title and href", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: `<ic-breadcrumb page-title="foo" href="/"></ic-breadcrumb>`,
    });

    expect(page.root).toMatchSnapshot(`
      <ic-breadcrumb href="/" page-title="foo" role="listitem">
        <mock:shadow-root>
          <div class="breadcrumb">
            <span aria-hidden="true" class="chevron">
              svg
            </span>
            <ic-link theme="inherit" class="breadcrumb-link" href="/">
              foo
            </ic-link>
          </div>
        </mock:shadow-root>
      </ic-breadcrumb>
    `);
  });

  it("should render with page-title, href, current", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: `<ic-breadcrumb page-title="foo" href="/" current="true"></ic-breadcrumb>`,
    });

    expect(page.root).toMatchSnapshot(`
      <ic-breadcrumb current="true" href="/" page-title="foo" role="listitem" aria-current="page">
        <mock:shadow-root>
          <div class="breadcrumb">
            <span aria-hidden="true" class="chevron">
              svg
            </span>
            <span class="current-page-container">
              foo
            </span>
          </div>
        </mock:shadow-root>
      </ic-breadcrumb>
    `);
  });

  it("should render with icon", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: `
      <ic-breadcrumb page-title="foo" href="/" current="true">
        <svg
          slot="icon"
          width="24"
          height="24"
          viewBox="0 0 24 24"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 6.19L17 10.69V18.5H15V12.5H9V18.5H7V10.69L12 6.19ZM12 3.5L2 12.5H5V20.5H11V14.5H13V20.5H19V12.5H22L12 3.5Z"
            fill="currentColor"
          ></path>
        </svg>
      </ic-breadcrumb>`,
    });

    expect(page.root).toMatchSnapshot(`
      <ic-breadcrumb current="true" href="/" page-title="foo" aria-current="page" role="listitem">
        <mock:shadow-root>
          <div class="breadcrumb">
            <span aria-hidden="true" class="chevron">
              svg
            </span>
            <span class="current-page-container">
              <slot name="icon"></slot>
              foo
            </span>
          </div>
        </mock:shadow-root>
        <svg fill="none" height="24" slot="icon" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
          <path d="M12 6.19L17 10.69V18.5H15V12.5H9V18.5H7V10.69L12 6.19ZM12 3.5L2 12.5H5V20.5H11V14.5H13V20.5H19V12.5H22L12 3.5Z" fill="currentColor"></path>
        </svg>
      </ic-breadcrumb>
    `);
  });

  it("should add the 'current-page' class when current prop is true", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: "<ic-breadcrumb id='ic-breadcrumb' current=true><ic-link href='/'>Link</ic-link></ic-breadcrumb>",
    });

    await page.rootInstance.setSlottedCurrentPageClass();

    expect(
      document
        .querySelector("ic-breadcrumb")
        ?.classList.contains("current-page")
    ).toBeTruthy;
  });

  it("should call 'setFocus' when breadcrumb is focused", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: `<ic-breadcrumb></ic-breadcrumb>`,
    });

    //Can't expect anything in this test - this is to increase code coverage only
    await page.rootInstance.setFocus().toHaveBeenCalled;
  });

  it("should prevent focus on the slotted link within the current breadcrumb", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: "<ic-breadcrumb id='ic-breadcrumb' current=true><ic-link href='/'>Link</ic-link></ic-breadcrumb>",
    });

    page.root?.setAttribute("current", "true");
    page.rootInstance.linkSlotContent.tabIndex = 0;
    await page.waitForChanges();

    // Cannot check for actual slot updates in the HTML due to assignedElements not being available in test environment
    const slottedLinkEl = document.createElement("a");
    page.rootInstance.slottedLinkEl = slottedLinkEl;
    jest
      .spyOn(page.rootInstance.slottedLinkEl, "removeAttribute")
      .mockImplementation();

    page.rootInstance.updatedSlottedLinkFocus();
    await page.waitForChanges();

    expect(page.rootInstance.linkSlotContent.tabIndex).toBe(-1);
    expect(page.rootInstance.slottedLinkEl.removeAttribute).toHaveBeenCalled();
  });

  it("should change to allow focus on slotted link when current prop is removed", async () => {
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: "<ic-breadcrumb id='ic-breadcrumb' current=true><ic-link href='/'>Link</ic-link></ic-breadcrumb>",
    });

    const slottedLinkEl = document.createElement("a");
    page.rootInstance.slottedLinkEl = slottedLinkEl;
    page.rootInstance.slottedLinkHref = "/";
    await page.waitForChanges();

    jest
      .spyOn(page.rootInstance.slottedLinkEl, "setAttribute")
      .mockImplementation();

    page.root?.setAttribute("current", "true");
    await page.waitForChanges();

    page.root?.setAttribute("current", "false");
    page.rootInstance.linkSlotContent.tabIndex = -1;
    page.rootInstance.updatedSlottedLinkFocus();
    await page.waitForChanges();

    expect(page.rootInstance.linkSlotContent.tabIndex).toBe(0);
    expect(page.rootInstance.slottedLinkEl.setAttribute).toHaveBeenCalled();

    page.rootInstance.linkSlotContent = null;
    await page.waitForChanges();

    expect(page.rootInstance.getSlottedLinkEl()).toBe(null);
  });
});
