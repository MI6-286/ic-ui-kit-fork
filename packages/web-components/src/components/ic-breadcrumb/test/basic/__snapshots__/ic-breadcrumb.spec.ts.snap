// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-breadcrumb should render with icon: 
      <ic-breadcrumb current="true" href="/" page-title="foo" aria-current="page" role="listitem">
        <mock:shadow-root>
          <div class="breadcrumb">
            <span aria-hidden="true" class="chevron">
              svg
            </span>
            <span class="current-page-container">
              <slot name="icon"></slot>
              foo
            </span>
          </div>
        </mock:shadow-root>
        <svg fill="none" height="24" slot="icon" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
          <path d="M12 6.19L17 10.69V18.5H15V12.5H9V18.5H7V10.69L12 6.19ZM12 3.5L2 12.5H5V20.5H11V14.5H13V20.5H19V12.5H22L12 3.5Z" fill="currentColor"></path>
        </svg>
      </ic-breadcrumb>
     1`] = `
<ic-breadcrumb aria-current="page" current="true" href="/" page-title="foo" role="listitem">
  <mock:shadow-root>
    <div class="breadcrumb">
      <span aria-hidden="true" class="chevron">
        svg
      </span>
      <span class="current-page-container">
        <slot name="icon"></slot>
        foo
      </span>
    </div>
  </mock:shadow-root>
  <svg fill="none" height="24" slot="icon" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
    <path d="M12 6.19L17 10.69V18.5H15V12.5H9V18.5H7V10.69L12 6.19ZM12 3.5L2 12.5H5V20.5H11V14.5H13V20.5H19V12.5H22L12 3.5Z" fill="currentColor"></path>
  </svg>
</ic-breadcrumb>
`;

exports[`ic-breadcrumb should render with page-title and href: 
      <ic-breadcrumb href="/" page-title="foo" role="listitem">
        <mock:shadow-root>
          <div class="breadcrumb">
            <span aria-hidden="true" class="chevron">
              svg
            </span>
            <ic-link theme="inherit" class="breadcrumb-link" href="/">
              foo
            </ic-link>
          </div>
        </mock:shadow-root>
      </ic-breadcrumb>
     1`] = `
<ic-breadcrumb href="/" page-title="foo" role="listitem">
  <mock:shadow-root>
    <div class="breadcrumb">
      <span aria-hidden="true" class="chevron">
        svg
      </span>
      <ic-link class="breadcrumb-link" href="/" theme="inherit">
        foo
      </ic-link>
    </div>
  </mock:shadow-root>
</ic-breadcrumb>
`;

exports[`ic-breadcrumb should render with page-title, href, current: 
      <ic-breadcrumb current="true" href="/" page-title="foo" role="listitem" aria-current="page">
        <mock:shadow-root>
          <div class="breadcrumb">
            <span aria-hidden="true" class="chevron">
              svg
            </span>
            <span class="current-page-container">
              foo
            </span>
          </div>
        </mock:shadow-root>
      </ic-breadcrumb>
     1`] = `
<ic-breadcrumb aria-current="page" current="true" href="/" page-title="foo" role="listitem">
  <mock:shadow-root>
    <div class="breadcrumb">
      <span aria-hidden="true" class="chevron">
        svg
      </span>
      <span class="current-page-container">
        foo
      </span>
    </div>
  </mock:shadow-root>
</ic-breadcrumb>
`;

exports[`ic-breadcrumb should render: 
      <ic-breadcrumb role="listitem">
        <mock:shadow-root>
          <div class="breadcrumb">
            <span aria-hidden="true" class="chevron">
              svg
            </span>
            <span class="link-wrapper" tabindex="0">
              <slot></slot>
            </span>
          </div>
        </mock:shadow-root>
      </ic-breadcrumb>
     1`] = `
<ic-breadcrumb role="listitem">
  <mock:shadow-root>
    <div class="breadcrumb">
      <span aria-hidden="true" class="chevron">
        svg
      </span>
      <div class="slotted-link-container">
        <span class="link-wrapper" tabindex="0">
          <slot></slot>
        </span>
      </div>
    </div>
  </mock:shadow-root>
</ic-breadcrumb>
`;
